File: src/automata/step.ts
import type { AutomataConfig, StateArray } from "./config";
import { createRuleResolver } from "./rules";

function automataStep0(
  config: AutomataConfig,
  resolver: (parents: ArrayLike<number>) => number,
  world: StateArray,
  row: number,
): StateArray {
  const padLeft = config.padLeft[(row - 1) % config.padLeft.length];
  const padRight = config.padRight[(row - 1) % config.padRight.length];
  const midLen = world.length - config.numParents + 1;
  const result = new Uint8Array(padLeft.length + midLen + padRight.length);
  result.set(padLeft, 0);
  for (let i = 0; i < midLen; i++) {
    result[padLeft.length + i] = resolver(world.subarray(i, i + config.numParents));
  }
  result.set(padRight, padLeft.length + midLen);
  return result;
}

export function automataStep(config: AutomataConfig, world: StateArray, row: number = 1): StateArray {
  const resolver = createRuleResolver(config.ruleMode, config.numParents, config.numStates, config.rules);
  return automataStep0(config, resolver, world, row);
}

export function automataStepBatch(
  config: AutomataConfig,
  initial: StateArray,
  stepCount: number,
  startRow: number = 1,
): StateArray[] {
  const result: StateArray[] = [];
  let current = initial;
  let row = startRow;
  const resolver = createRuleResolver(config.ruleMode, config.numParents, config.numStates, config.rules);
  while (result.length < stepCount) {
    current = automataStep0(config, resolver, current, row);
    result.push(current);
    row++;
  }
  return result;
}

File: rust-lib/src/step.rs
use crate::parser::AutomataConfig;
use crate::packed::PackedRow;
use crate::rules::resolve;
use crate::mods::apply_mods;

pub fn automata_step(
    config: &AutomataConfig,
    lookup: &[u8],
    world: &PackedRow,
    row: usize,
) -> PackedRow {
    let n_frames = config.pad_left.len().max(1);
    let frame_idx = (row - 1) % n_frames;
    let pad_l = &config.pad_left[frame_idx % config.pad_left.len()];
    let pad_r = &config.pad_right[frame_idx % config.pad_right.len()];

    let mid_len = world.len.saturating_sub(config.num_parents - 1);
    let out_len = pad_l.len + mid_len + pad_r.len;
    let mut result = PackedRow::new(out_len, config.bpc);

    for i in 0..pad_l.len {
        result.set(i, pad_l.get(i));
    }

    let mut parents = vec![0u8; config.num_parents];
    for i in 0..mid_len {
        for p in 0..config.num_parents {
            parents[p] = world.get(i + p);
        }
        let state = resolve(lookup, &parents, config.rule_mode, config.num_states);
        result.set(pad_l.len + i, state);
    }

    for i in 0..pad_r.len {
        result.set(pad_l.len + mid_len + i, pad_r.get(i));
    }

    apply_mods(&mut result, row, &config.mods);
    result
}

pub fn make_initial_row(config: &AutomataConfig) -> PackedRow {
    let mut initial = PackedRow::from_slice(
        &(0..config.initial.len).map(|i| config.initial.get(i)).collect::<Vec<_>>(),
        config.bpc,
    );
    apply_mods(&mut initial, 0, &config.mods);
    initial
}
